File: src/algorithms/trees/properties/cousins-in-binary-tree/cousins-in-binary-tree.test.ts
import { describe, it, expect } from "vitest";
import { cousinsInBinaryTree } from "./sources/cousins-in-binary-tree.ts?fn";

interface BSTNode {
  value: number;
  left: BSTNode | null;
  right: BSTNode | null;
}
function node(value: number, left: BSTNode | null = null, right: BSTNode | null = null): BSTNode {
  return { value, left, right };
}

describe("cousinsInBinaryTree", () => {
  it("returns true for nodes 1 and 5 (cousins in 7-node BST)", () => {
    const root = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));
    expect(cousinsInBinaryTree(root, 1, 5)).toBe(true);
  });

  it("returns false for siblings (same parent)", () => {
    const root = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));
    expect(cousinsInBinaryTree(root, 1, 3)).toBe(false);
  });

  it("returns false for nodes at different depths", () => {
    const root = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));
    expect(cousinsInBinaryTree(root, 2, 1)).toBe(false);
  });

  it("returns false for null root", () => {
    expect(cousinsInBinaryTree(null, 1, 2)).toBe(false);
  });

  it("returns true for another valid cousin pair (3 and 7)", () => {
    const root = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));
    expect(cousinsInBinaryTree(root, 3, 7)).toBe(true);
  });
});

File: src/algorithms/trees/properties/cousins-in-binary-tree/CousinsInBinaryTree_test.cpp
// g++ -o cousins_test CousinsInBinaryTree_test.cpp && ./cousins_test
#include "sources/CousinsInBinaryTree.cpp"
#include <cassert>
#include <iostream>

TreeNode* makeCIBTNode(int value, TreeNode* left = nullptr, TreeNode* right = nullptr) {
    TreeNode* node = new TreeNode(value);
    node->left = left;
    node->right = right;
    return node;
}

TreeNode* buildCIBT7NodeTree() {
    return makeCIBTNode(4,
        makeCIBTNode(2, makeCIBTNode(1), makeCIBTNode(3)),
        makeCIBTNode(6, makeCIBTNode(5), makeCIBTNode(7)));
}

int main() {
    // test: cousins 1 and 5
    assert(cousinsInBinaryTree(buildCIBT7NodeTree(), 1, 5) == true);

    // test: siblings not cousins
    assert(cousinsInBinaryTree(buildCIBT7NodeTree(), 1, 3) == false);

    // test: different depths not cousins
    assert(cousinsInBinaryTree(buildCIBT7NodeTree(), 2, 1) == false);

    // test: null root returns false
    assert(cousinsInBinaryTree(nullptr, 1, 2) == false);

    // test: cousins 3 and 7
    assert(cousinsInBinaryTree(buildCIBT7NodeTree(), 3, 7) == true);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
